sys/senml/senml.c

```diff
@@ -0,0 +1,357 @@
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_UNIT)) {
+                    pack->base_info->base_unit = tmp_val;
+                    found_base_info = true;
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_VALUE)) {
+                    pack->base_info->base_value = atof(tmp_val);
+                    found_base_info = true;
+                }
+            }
+
+            if (!strcmp(tmp_key, SJ_NAME)) {
+                (pack->records)[array_num].name = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_UNIT)) {
+                (pack->records)[array_num].unit = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_FLOAT;
+                (pack->records)[array_num].value.f = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_STRING_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_STRING;
+                (pack->records)[array_num].value.s = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_BOOL_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BOOL;
+                if (!strcmp(tmp_val, "true")) {
+                    (pack->records)[array_num].value.b = true;
+                }
+                else if (!strcmp(tmp_val, "false")) {
+                    (pack->records)[array_num].value.b = false;
+                }
+            }
+            else if (!strcmp(tmp_key, SJ_DATA_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BINARY;
+                (pack->records)[array_num].value.d = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE_SUM)) {
+                (pack->records)[array_num].value_sum = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_TIME)) {
+                (pack->records)[array_num].time = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_UPDATE_TIME)) {
+                (pack->records)[array_num].update_time = atof(tmp_val);
+            }
+        }
+    }
+    
+    if (!found_base_info) {
+        pack->base_info = NULL;
+    }
+
+    pack->num = (array_num + 1);
+
+    return 0;
+}
+
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    if (len < 3) {
+        DEBUG("ERROR: cannot encode pack to JSON because output buffer is too small\n");
+        return -1;
+    }
+
+    // using len - 3 for content to leave space for closing } and ] and '\0'
+    ssize_t cnt_len = len - 3;
+
+    size_t insert_pos = 0;
+
+    output[0] = '[';
+    insert_pos++;
+    cnt_len--;
+
+    int chars_written;
+
+    if (!pack->base_info) {
+        goto encode_records;
+    }
+
+    output[insert_pos] = '{';
+    insert_pos++;
+    cnt_len--;
+
+    if (pack->base_info->version) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bver\":%u,",
```

Comment:
If `len == 4`, `cnt_len == -1` here. `cnt_len` will be casted to `size_t` because `snprintf` receives `size_t`, results in underflow.